Fixes for documentation search

megalobyte opened the pull request:
Fixes #44991
*Bugsquad edit:* Fixes #32687

Fixes a couple of issues when using `FindBar`:
1. Find next result
2. Find previous result
3. Wrap around
4. Update the scroll bar so that the next result is displayed (not centered, could be improved)  
5. `FindBar` does not use sensitive matches, so use case insensitive search when counting results

I used `Transform2D` as the sample documentation for testing.

megalobyte commented:
Added clang format. The search can now find multiple results in the same line. It's still unable to find multiple results in a table. 

I'm tempted to revert the changes that were introduced with the RichTextLabel refactor. It makes more sense to me to iterate through the different items, rather than by each line. Any ideas as to what this search method supports that the previous method didn't?

megalobyte commented:
This branch isn't currently ready for review. I found another bug with the previous search, but I have a fix for it on my local branch. Also, I may have a fix for the table search. I'm going to try to implement it when I get home

pycbouh commented:
You can mark your PR as draft in case you want to work a bit more on it.

megalobyte commented:
I think I got everything working. At the very least, none of the edge cases I came across had any issues :sweat:

megalobyte commented:
By the way, is there a story for `Shift+Enter` not invoking previous search? I know the source of the problem. I just don't know if I should add `Shift+Enter` to `ui_text_newline`, make a new input event, or add it to a different input event that I overlooked.

akien-mga commented:
> By the way, is there a story for `Shift+Enter` not invoking previous search? I know the source of the problem. I just don't know if I should add `Shift+Enter` to `ui_text_newline`, make a new input event, or add it to a different input event that I overlooked.

There should probably be `ui_text_search_next` and `ui_text_search_previous` shortcuts. Using `ui_text_newline` for "Search Next" is probably wrong in the first place.

akien-mga commented on scene/gui/rich_text_label.cpp:
```diff
@@ -3511,23 +3533,81 @@ bool RichTextLabel::_search_line(ItemFrame *p_frame, int p_line, const String &p
+			scroll_to_line(selection.from_frame->line + selection.from_line);
+			update();
+			return true;
+		};
```
Unnecessary semicolon.

megalobyte commented:
> > By the way, is there a story for `Shift+Enter` not invoking previous search? I know the source of the problem. I just don't know if I should add `Shift+Enter` to `ui_text_newline`, make a new input event, or add it to a different input event that I overlooked.
> 
> There should probably be `ui_text_search_next` and `ui_text_search_previous` shortcuts. Using `ui_text_newline` for "Search Next" is probably wrong in the first place.

Sounds good. The changes were pretty minimal, so I took the liberty of adding them to this PR since adding them seemed relevant enough

Paulb23 commented on scene/gui/line_edit.cpp:
```diff
@@ -357,8 +357,8 @@ void LineEdit::_gui_input(Ref<InputEvent> p_event) {
-		// Default is ENTER, KP_ENTER. Cannot use ui_accept as default includes SPACE
-		if (k->is_action("ui_text_newline", true)) {
+		// Default is ENTER. Cannot use ui_accept as default includes SPACE
+		if (k->is_action("ui_text_search_next", true) || k->is_action("ui_text_search_previous", true)) {
```
This is tricky, as search is misleading.

I'm thinking about `ui_text_submit` and renaming the signal to `text_submitted` so they align.

megalobyte replied:
What are your thoughts about naming it to `ui_text_entered`? I'm a bit hesitant to rename the signal to `text_submitted`, since a lot of other functions are connecting to `text_entered`

Paulb23 replied:
Yeah the problem with `ui_text_entered` is that it somewhat implies that text has been entered, rather then the `enter` key has been pressed. Given that we're making it a configurable shortcut, it may not event be the `enter` key, so would be even more confusing.

megalobyte replied:
Agreed, it's definitely not ideal. So you think I should rename the signal from `text_entered` to `text_submitted` and update all the files that use `LineEdit`? Or would it be better to just add a new signal called `text_submitted`?

Paulb23 replied:
Go for the rename, this is for 4.0 only so it's okay to break compat. 

Though probably worth doing the rename as a separate commit.

megalobyte replied:
Okay. I will keep the signal name `text_entered` for this PR. I will mark this comment resolved once I submit a follow on PR that renames `text_entered` to `text_submitted`

megalobyte replied:
Forget the extra PR. I forgot we can squash the commit after the review if everything gets approved

Paulb23 reviewed (changes requested):
Overall LGTM.

Minor nitpick, as I'm not sure on the new shortcut naming as `LineEdit` is a generic node.

Paulb23 commented on scene/gui/rich_text_label.h:
```diff
@@ -392,7 +392,8 @@ class RichTextLabel : public Control {
 
 	String _get_line_text(ItemFrame *p_frame, int p_line, Selection p_sel) const;
-	bool _search_line(ItemFrame *p_frame, int p_line, const String &p_string, Item *p_from, Item *p_to);
+	bool _search_line(ItemFrame *p_frame, int p_line, const String &p_string, int char_idx, bool reverse_search);
```
`char_idx` and `reverse_search` need to be preficed with `p_` ditto for `_search_table`

megalobyte replied:
I thought the `p_` prefix is used to denote a pointer. Does `p_` mean something else in godot?

Paulb23 commented on scene/gui/rich_text_label.cpp:
```diff
@@ -3511,23 +3533,81 @@ bool RichTextLabel::_search_line(ItemFrame *p_frame, int p_line, const String &p
+		// Next, check to see if the current search result is in a table
+		if (selection.from_frame->parent != nullptr && selection.from_frame->parent->type == ITEM_TABLE) {
+			// Find last search result in table
+			ItemTable *p_table = static_cast<ItemTable *>(selection.from_frame->parent);
```
`p_` is for params so needs renaming to `parent_table`? same for `p_element`

megalobyte replied:
This should be fixed with the latest push

EricEzaM commented on doc/classes/ProjectSettings.xml:
```diff
@@ -733,6 +733,10 @@
 		</member>
 		<member name="input/ui_text_scroll_up.OSX" type="Dictionary" setter="" getter="">
 		</member>
+		<member name="input/ui_text_search_next" type="Dictionary" setter="" getter="">
+		</member>
+		<member name="input/ui_text_search_previous" type="Dictionary" setter="" getter="">
+		</member>
```
How come these are added to the doco? I can't find references to them anywhere else.

megalobyte replied:
These shortcuts were added in a previous commit. I renamed them to address feedback, but forgot to run the doctool.

EricEzaM commented on core/input/input_map.cpp:
```diff
@@ -667,6 +668,11 @@ const OrderedHashMap<String, List<Ref<InputEvent>>> &InputMap::get_builtins() {
 	inputs.push_back(InputEventKey::create_reference(KEY_MENU));
 	default_builtin_cache.insert("ui_menu", inputs);
 
+	inputs = List<Ref<InputEvent>>();
+	inputs.push_back(InputEventKey::create_reference(KEY_ENTER));
+	inputs.push_back(InputEventKey::create_reference(KEY_ENTER | KEY_MASK_SHIFT));
+	default_builtin_cache.insert("ui_text_enter", inputs);
+
```
It seems wrong to have Shift+Enter for the "Text Entered" shortcut...

I think a better solution could be to only have `KEY_ENTER` and `KEY_KP_ENTER` assigned to this, and in the `is_action` check in LineEdit do a non-exact check (i.e. second param is `false`). This would allow for Shift-Enter to trigger the signal, and then the current logic downstream (checking if Shift is pressed) would still work as it currently does.

megalobyte replied:
That is good to know! I'm still learning how godot works, but I should have looked more closely at how `is_action` works. Will be updated in latest push

EricEzaM commented on scene/gui/line_edit.cpp:
```diff
@@ -357,8 +357,8 @@ void LineEdit::_gui_input(Ref<InputEvent> p_event) {
 
-		// Default is ENTER, KP_ENTER. Cannot use ui_accept as default includes SPACE
-		if (k->is_action("ui_text_newline", true)) {
+		// Default is ENTER. Cannot use ui_accept as default includes SPACE
+		if (k->is_action("ui_text_enter", true)) {
```
```suggestion
		// Default is ENTER and KP_ENTER. Cannot use ui_accept as default includes SPACE
		if (k->is_action("ui_text_enter", false)) {
```

This is the change I am talking about it my comment above.

megalobyte replied:
Added in latest commit

Paulb23 commented on editor/animation_track_editor.cpp:
```diff
@@ -2725,7 +2725,7 @@ void AnimationTrackEdit::_gui_input(const Ref<InputEvent> &p_event) {
 			path_popup->add_child(path);
 			path->set_anchors_and_offsets_preset(PRESET_WIDE);
-			path->connect("text_entered", callable_mp(this, &AnimationTrackEdit::_path_entered));
+			path->connect("text_submitted", callable_mp(this, &AnimationTrackEdit::_path_entered));
```
Missed `_path_entered` to `_path_submitted`

megalobyte replied:
Good catch! I should have paid more attention to my exclusion search filter :facepalm: 

Updated `expression.xml` and `animation_track_editor.cpp`. I've also squashed the commits into one, but I saved a branch locally that preserves the separate commits in case some one wants to review the commits separately

Paulb23 reviewed (commented):
Couple small misses of `text_entered` in docs `expression.xml`, and one inline.

After that LGTM.

megalobyte committed:
Documentation search fixes

Updates rich_text_label so that the built-in documentation can be searched
Previously, it would only find the first result and would not select other results
Renames "_entered" functions to "_submitted"

akien-mga commented:
FYI, the email used to author your commit doesn't seem to be linked to your GitHub account (see header of https://patch-diff.githubusercontent.com/raw/godotengine/godot/pull/49258.patch). It's not a problem for Git or GitHub, but if you want to claim ownership of the commit for street creds, you can add it as secondary email in your profile.

megalobyte commented:
@akien-mga Thanks. I thought I updated my email address, but I guess not. I'll update it when I get home. Please postpone merging until then

EricEzaM reviewed (approved):
Changes I suggested were incorporated, they look good.

akien-mga commented:
Thanks!

> I'll update it when I get home. Please postpone merging until then

Oops, sorry about that. Merged while on phone and I just recalled this after hitting the button.

megalobyte commented:
No worries. I just added the email address to my github account like you suggested. Thanks
